File: Data Structures & Algorithms/two-integer-sum-ii/submission-0.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        /*
        the brute force solution is o(n^2) which means for each value
        we check every other possible value to see if it adds up to target

        optimal solution is to use two pointers with a lp at 1
        and the rp num.length as it is one indexed.


        */
        
    

    int lp = 0;
    int rp = numbers.length-1;

    while(lp<rp){
        if(numbers[lp]+numbers[rp]>target){
            rp--;
        }else if(numbers[lp] + numbers[rp]<target){
            lp++;
        }else{
            //return the index of lp and rp
            return new int[]{lp+1,rp+1};
        }

    }
    return new int[]{};
    }
}

File: Data Structures & Algorithms/two-integer-sum/submission-0.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        

        """

        [3,4,1,2]. target 5

        we need to be able to find a pair of indices that equal to the target
        but instead of the brute force solution wnhere we compare each value with every other value,
        we can iterate over each array and check if we have already counted it (in our hashmap)

        {3:1 4,2}

        as i know the target, is 5 for instance
        at each value i can check if target - nums[i] exists in my map
        so the key is the index and the value is the value at the iterable
        
        """
        
        X = {} #stores the numbers as the key and the index as the value

        for i in range(len(nums)):
            
            if target - nums[i] in X:
                return [X[target-nums[i]], i]
            else:
                X[nums[i]] = i    

                
                    
    



    



